theano/tensor/nnet/abstract_conv.py

```diff
@@ -465,10 +505,30 @@ def make_node(self, img, kern):
         return Apply(self, [img, kern], [output])
 
     def perform(self, node, inp, out_):
-        raise NotImplementedError(
-            'AbstractConv2d theano optimization failed. '
-            'Did you exclude both "conv_dnn" and "conv_gemm" from '
-            'the optimizer? Is cudnn available and does the GPU support it?')
+        img, kern = inp
+        img = numpy.asarray(img)
+        kern = numpy.asarray(kern)
+        o, = out_
+        mode = self.border_mode
+
+        if mode == "full":
+            mode = (kern.shape[2] - 1, kern.shape[3] - 1)
+        elif mode == "half":
+            mode = (kern.shape[2] // 2, kern.shape[3] // 2)
+        if isinstance(mode, tuple):
+            pad_h, pad_w = map(int, mode)
+            mode = "valid"
+            new_img = numpy.zeros((img.shape[0], img.shape[1],
+                                   img.shape[2] + 2 * pad_h,
+                                   img.shape[3] + 2 * pad_w), dtype=img.dtype)
+            new_img[:, :, pad_h:img.shape[2] + pad_h, pad_w:img.shape[3] + pad_w] = img
+            img = new_img
+        if not self.filter_flip:
+            kern = kern[:, :, ::-1, ::-1]
+        conv_out = self.corr2d(img, kern, mode)
```

Comment:
At this point, `mode` is always "valid", right?
So maybe we can just get rid of that keyword here, or call "valid" explicitly.
Also, we should check that `mode` is either "full", "half", "valid" or a pair of int, because we may accept new strings in the future.